### PR TITLE
TVPType. Add support custom user tag

### PR DESCRIPTION
**Pull request by NikitaDef**
By the tag, TVPType can omit the fields. Which value is "-".
This feature is useful when users want to use part of fields in struct.

### CODE REVIEW

**Comment by kardianos** [comment]
The code looks reasonable, but I haven't been following the TVP (sorry). I'm not seeing much docs. Can I see some docs on what struct tags are accepted. For the tag names, I might expect something like:
```
type X struct {
    F1 string `tvp:"-"`
    F2 string `tvp:"f2,skip"`
}
```

I just don't think I completly understand the intent.

**Comment by NikitaDef** [comment]
@kardianos Example about behavior:
```
type X struct {
	F1 string `tvp:"-"` //driver skips field because 'TVPCustomTag' is 'tvp'
	F2 string `tvp:"f2,skip"`
	F3 string `tvp:"f2,skip"`
	F4 string `json:"-"` 
}
mssql.TVPType{
	TVPCustomTag:  "tvp",
}
```

Also [Documentaion](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/0dfc5367-a388-4c92-9ba4-4d28e775acbc)
ColName MUST be a zero-length string in the TVP. 

And I think driver can support only tag value "-", because others have the same behavior.

**Comment by kardianos** [comment]
Okay, so this PR does two things: it allows setting a custom tag name with TVPCustomTag, then allows setting names or properties to these tags.

My initial reaction is just to use a standard tag name for setting field names or skipping them, rather then make the tag name customization. (You can't set the tag name for json or other packages for example). Can you explain why you want a custom tag name, rather then just choosing a tag name like "tvp"?

**Comment by NikitaDef** [comment]
@kardianos I get limit when I use mssql.TVPType with protobuf. 
With each generated struct I get 3 additional fields. 
I mustn't remove them or add 'tvp' tag because after regeneration I lose the changes. 
And I must have to create a copy of struct without 3 additional fields.
And TVPCustomTag helps me to ignore them. 

Example: 3 additional fields protobuf.
```
	XXX_NoUnkeyedLiteral struct{}     `json:"-"`
	XXX_unrecognized     []byte       `json:"-"`
	XXX_sizecache        int32        `json:"-"`
```

**Comment by kardianos** [comment]
@NikitaDef Thank you. I now understand your situation and motivation. I'm going to read it over once more with that in mind. Thanks.

**Comment by NikitaDef** [comment]
@kardianos are you going to merge it? 
Or do you have another idea about this situation?

**Comment by kardianos** [comment]
@NikitaDef Sorry, this got lost. For the time being, what if we looked at both a json and tvp tag. If json omits it, also omit it unless the tvp tag gives it a name. For instance:
```
type T struct {
    V1 string `json:"-" tvp:"v1"` // This would be marshaled as field v1.
    XXX string `json:"-"` // This would be ignored.
}
```

Maybe this would get you to where you want to go without adding yet another config option to each struct, but still being generally useful. I think the protobuf situation is a great motivation for something to be done here.

**Comment by NikitaDef** [comment]
@kardianos 'tvp' is the highest priority and default tag.
For instance behavior:
```
//when TVPCustomTag is "json:",
type T struct {
    V1 string `json:"v1" tvp:"v1"` // This would be marshaled as field v1.
    V2 string `json:"-" tvp:"v2"` // This would be marshaled as field v2.
    V2 string `json:"v2" tvp:"-"` // This would be ignored.
    V3 string `json:"-" tvp:"-"` // This would be ignored.

    V3 string `json:"v3"` // This would be marshaled as field v3.
    V4 string `json:"-"` // This would be ignored.
}

//when TVPCustomTag is "",
type T struct {
    V1 string `json:"v1" tvp:"v1"` // This would be marshaled as field v1.
    V2 string `json:"-" tvp:"v2"` // This would be marshaled as field v2.
    V3 string `json:"v3" tvp:"-"` // This would be ignored.
    V4 string `json:"-" tvp:"-"` // This would be ignored.

    V5 string `json:"v5"` // This would be marshaled as field v5.
    V6 string `json:"-"` // This would be marshaled as field v6..
}
```
Do I understand correctly about behavior?

**Comment by kardianos** [comment]
Yeah, except I'm thinking just get rid of TVPCustomTag and just look at the json flag by default, so only your top section would apply (as if TVPCustomTag is always "json". Unless you can think of a concrete situation where that would be undesirable or insufficient.

**Comment by NikitaDef** [comment]
@kardianos Sorry, I misunderstood your previous comment.
I agree, I'll remove TVPCustomTag

**Comment by NikitaDef** [comment]
@kardianos @denisenkom I did some changes.
Also I did changes in TVPType #458.

File: tvp_go19.go [hunk 1]
```diff
@@ -8,29 +8,52 @@ import (
...
 var (
 	ErrorEmptyTVPName        = errors.New("TVPTypeName must not be empty")
 	ErrorTVPTypeSlice        = errors.New("TVPType must be slice type")
 	ErrorTVPTypeSliceIsEmpty = errors.New("TVPType mustn't be null value")
+	ErrorTVPSkip             = errors.New("all fields mustn't skip")
+	ErrorTVPObjectName       = errors.New("wrong tvp name")
 )
 
 //TVPType is driver type, which allows supporting Table Valued Parameters (TVP) in SQL Server
 type TVPType struct {
```
[the remark this answers is not in the document]
Comment (NikitaDef) [reply tvp_go19.go]: I'll remove prefixes

File: tvp_go19.go [hunk 2]
```diff
@@ -8,29 +8,52 @@ import (
...
 	//TVP param name, mustn't be default value
 	TVPTypeName string
-	//TVP scheme name
-	TVPScheme string
-	//TVP Value. Param must be the slice, mustn't be nil
+	//TVP Value Param must be the slice, mustn't be nil
 	TVPValue interface{}
+	//TVPCustomTag If the field tag is "-", the field is always omit
+	tvpCustomTag string
```
[the remark this answers is not in the document]
Comment (NikitaDef) [reply tvp_go19.go]: I'll remove tvpCustomTag

File: tvp_go19_test.go [hunk 3]
```diff
@@ -160,6 +268,42 @@ func TestTVPType_check(t *testing.T) {
 	}
 }
 
+func TestTVPType_encode(t *testing.T) {
+	type fields struct {
+		TVPTypeName  string
+		TVPScheme    string
+		TVPValue     interface{}
+		TVPCustomTag string
```
[the remark this answers is not in the document]
Comment (NikitaDef) [reply tvp_go19_test.go]: I'll remove TVPCustomTag

File: tvp_go19.go [hunk 4]
```diff
@@ -8,29 +8,52 @@ import (
...
 		return ErrorEmptyTVPName
 	}
+	if !isProc(tvp.TVPTypeName) {
+		return ErrorEmptyTVPName
+	}
+	schema, name, err := getSchemeAndName(tvp.TVPTypeName)
+	if err != nil {
+		return err
+	}
+	tvp.TVPTypeName = name
```
[the remark this answers is not in the document]
Comment (NikitaDef) [reply tvp_go19.go]: By [documentation](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/0dfc5367-a388-4c92-9ba4-4d28e775acbc) Schema and TypeName are different fields.
And I don't know how I can avoid modifying the TVPType.

```
TVP_TYPENAME definition

 DBNAME        =   B_VARCHAR ; Database where TVP type resides
 OwningSchema  =   B_VARCHAR ; Schema where TVP type resides
 TypeName      =   B_VARCHAR ; TVP type name
 TVP_TYPENAME  =   DbName
                   OwningSchema
                   TypeName
```

File: tvp_go19.go [hunk 5]
```diff
@@ -55,9 +78,8 @@ func (tvp TVPType) encode() ([]byte, error) {
 	if err != nil {
 		return nil, err
 	}
-	writeBVarChar(buf, tvp.TVPScheme)
+	writeBVarChar(buf, tvp.tvpScheme)
```
[the remark this answers is not in the document]
Comment (NikitaDef) [reply tvp_go19.go]: I'll call `getSchemeAndName` in `encode`

File: tvp_go19.go [hunk 6]
```diff
@@ -8,29 +8,52 @@ import (
...
-	TVPScheme string
-	//TVP Value. Param must be the slice, mustn't be nil
+	//TVP Value Param must be the slice, mustn't be nil
 	TVPValue interface{}
+	//TVPCustomTag If the field tag is "-", the field is always omit
+	tvpCustomTag string
+	//tvp scheme name
+	tvpScheme string
+	//tvpFieldIndexes
+	tvpFieldIndexes []int
```
[the remark this answers is not in the document]
Comment (NikitaDef) [reply tvp_go19.go]: I added `columnTypes` in tvpFieldIndexes at return param.
And `encode` doesn't need to calculate

File: tvp_go19_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (NikitaDef) [reply tvp_go19_test.go]: Removed

File: tvp_go19.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (NikitaDef) [reply tvp_go19.go]: Removed

File: tvp_go19.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (NikitaDef) [reply tvp_go19.go]: Removed prefixes

**Comment by NikitaDef** [comment]
Current `TVPType struct`  #458

```
//TVPType is driver type, which allows supporting Table Valued Parameters (TVP) in SQL Server
type TVPType struct {
	//TVP param name, mustn't be default value
	TVPTypeName string
	//TVP Value Param must be the slice, mustn't be nil
	TVPValue interface{}
}
```

File: tvp_go19.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (NikitaDef) [reply tvp_go19.go]: I changed to `encode(schema, name string)`
Params pass from `makeParamExtra`

**Pull request merged**